### PR TITLE
Add sorting for person attributes

### LINKED ISSUE #17: Add Sort By Priority
As an insurance agent, I can sort clients by priority, so that I can deal with client with higher priority status first

### PR DESCRIPTION
**Pull request by solomonng2001**
Unable to sort clients by their attributes.

Addition of sort will allow convenient access when looking up clients. Closes #17, #20

Let's
* make person attributes comparable, by implementing the interface.
* Add sorting command and parser.
* Add person comparator to abstract to logic for handling comparators.
* Add tests in model, commands and parser.
* Update ui to refresh client list after every command execute.

Notes on removing "final" for filteredPersons
* filteredPersons is designed to be immutable and final, with an inbuilt predicate filtering method only.
* With the introduction of sorting the people, this design is too restrictive on the functionality of the list.
* FilteredList does not support any sorting or amendments to the list other than filtering by a predicate.
* One alternative is to create a separate SortedList, which leads to more than 1 truth in terms of personLists, and complicates when to refer to either list throughout multiple commands. Additionally, when sortedList is displayed, this will affect the current implementation of multiple commands that reference clients by their index in filteredList.
* Another alternative is to expand the functionality of filteredPersons by changing the type to another List type. This is a much larger change compared to a measured change of retaining the FilteredList type but recreating the list to sort and reassigning to filteredPersons.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 1]
```diff
@@ -11,7 +11,7 @@
  * Represents a Client's birthday in the address book.
  * Guarantees: immutable; is valid as declared in {@link #isValidBirthday(String)}
  */
-public class Birthday {
+public class Birthday implements Comparable<Birthday> {
     public static final String MESSAGE_CONSTRAINTS = DateUtil.getMessageConstraintsForDateType("Birthday");
```
Comment (ReganChoy) [inline src/main/java/seedu/address/model/person/Birthday.java:15]: Good use of Comparable to help with sorting

File: src/main/java/seedu/address/model/person/PersonComparator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package seedu.address.model.person;
```
Comment (ReganChoy) [inline src/main/java/seedu/address/model/person/PersonComparator.java:1]: Good further abstraction by making another PersonComparator class

**Comment by ReganChoy** [comment]
Good use of the Comparable interface in all classes to support the sorting command.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -166,6 +167,13 @@ public void updateFilteredPersonList(Predicate<Person> predicate) {
         filteredPersons.setPredicate(predicate);
     }
 
+    @Override
+    public void sortFilteredPersonList(Comparator<Person> comparator) {
+        requireNonNull(comparator);
+        ObservableList<Person> sortedList = this.addressBook.getPersonList().sorted(comparator);
+        filteredPersons = new FilteredList<>(sortedList);
+    }
+
```
Comment (yorklim) [inline src/main/java/seedu/address/model/ModelManager.java:176]: The sort will cause the filtered list to reset and will return a full sorted list, is it possible to keep sort the already filtered list?

File: src/main/java/seedu/address/model/person/PersonComparator.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            comparator = Comparator.comparing(Person::getBirthday);
+            break;
+        case LASTMET:
+            comparator = Comparator.comparing(Person::getLastMet);
+            break;
+        case SCHEDULE:
+            comparator = Comparator.comparing(Person::getSchedule);
+            break;
+        default:
+            return Comparator.comparing(Person::getName);
```
Comment (yorklim) [inline src/main/java/seedu/address/model/person/PersonComparator.java:44]: Shouldn't the default sort return the list unmodified?

**Review by yorklim — CHANGES_REQUESTED** [review]
Overall LGTM! Just unsure if some of the implementations should be done that way.

File: src/main/java/seedu/address/model/person/PriorityValue.java [hunk 5]
```diff
@@ -8,8 +8,8 @@
 /**
  * Represents the value of a person's priority in the address book.
  */
-public enum PriorityValue {
-    LOW, MEDIUM, HIGH, VIP;
+public enum PriorityValue { // Already implements Comparable<PriorityValue> in the original code
+    LOW, MEDIUM, HIGH, VIP; // Order of priority values matters for compareTo method (low < medium < high < vip)
```
Comment (yorklim) [inline src/main/java/seedu/address/model/person/PriorityValue.java:12]: I like that you use an enum class to represent the different values, to ensure that no other incorrect values are taken. Good Job!

File: src/main/java/seedu/address/logic/commands/SortCommand.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package seedu.address.logic.commands;
```
Comment (yorklim) [inline src/main/java/seedu/address/logic/commands/SortCommand.java:1]: Nice and clean code! Good job with the Javadoc comments as well! Well Written

File: src/main/java/seedu/address/model/person/PersonComparator.java (hunk 2 above)
Comment (yorklim) [inline src/main/java/seedu/address/model/person/PersonComparator.java:1]: Should the comparator be in the person directory? Or can we consider putting it in the model instead, as having the comparator in the person directory seems like the comparator is part of the person.

File: src/main/java/seedu/address/model/person/SortOrder.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
+package seedu.address.model.person;
+
+/**
+ * Represents the sort order of the client list.
+ */
+public enum SortOrder {
```
Comment (getsquared) [inline src/main/java/seedu/address/model/person/SortOrder.java:6]: I like that you abstracted out SortCriteria and SortOrder into their own classes!

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 8]
```diff
@@ -207,6 +230,15 @@ private void refreshRemindersPanel() {
         addRemindersPanel();
     }
 
+    /**
+     * Refreshes the client view panel, client list panel and reminders panel.
+     */
+    private void refreshAllPanels() {
```
Comment (getsquared) [inline src/main/java/seedu/address/ui/MainWindow.java:236]: Nice thinking to refresh all panels at once, I did not think of that

File: src/test/java/seedu/address/logic/parser/SortCommandParserTest.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.Messages;
+import seedu.address.logic.commands.SortCommand;
+import seedu.address.model.person.SortCriteria;
+import seedu.address.model.person.SortOrder;
+
+public class SortCommandParserTest {
```
Comment (getsquared) [inline src/test/java/seedu/address/logic/parser/SortCommandParserTest.java:15]: Very comprehensive testing done here, keep up the good work 😄

**Review by getsquared — APPROVED** [review]

File: src/main/java/seedu/address/model/person/SortOrder.java (hunk 7 above)
Comment (getsquared) [inline src/main/java/seedu/address/model/person/SortOrder.java:6]: I like that you abstracted out SortCriteria and SortOrder!

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 8 above)
Comment (getsquared) [inline src/main/java/seedu/address/ui/MainWindow.java:236]: Great thinking on having a function to refresh all panels at once!

File: src/test/java/seedu/address/model/person/SortCriteriaTest.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
+package seedu.address.model.person;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+
+
+public class SortCriteriaTest {
```
Comment (getsquared) [inline src/test/java/seedu/address/model/person/SortCriteriaTest.java:10]: Very comprehensive testing done here, keep up the good work 😄

**Review by getsquared — COMMENTED** [review]
LGTM! 👍

**Pull request merged**